### ISSUE TITLE
Excel 2016: column headers not read when added using create table feature

### ISSUE THREAD

**Issue by leonardder** [nvaccess/nvda#8791]
### Steps to reproduce:
1. Open Excel 2016
2. Create a table with several rows and columns, where the first row contains headers
3. Select the range of cells
4. Press ctrl+l
5. Enable the check box that says that the table contains headers
6. Press ok
7. Navigate through the table

### Actual behavior:
NVDA does not announce column headers.

### Expected behavior:
NVDA announces column headers

#### NVDA Installed/portable/running from source:
Installed

#### NVDA version:
2018.3.1

#### Windows version:
Windows 10 1803

#### Name and version of other software in use when reproducing the issue:
Excel 16.0.10730.20102

#### Other information about your system:
When using JAWS 2018, JAWS correctly reads headers created using this method. Therefore, I assume that there ought to be a way to identify headers as such.

#### Does the issue still occur after restarting your PC?
Yes

#### Have you tried any other versions of NVDA?
No

**Comment by CyrilleB79** [comment]
Two things:
1. @leonardder the title reads "Excel 2016: column and row headers not read when added using create table feature", but in the description, you only speak of column headers. Can you confirm that row headers are out of the scope of this issue and that there is no way to create native row headers with Excel?

2. In 2023.1 and some other recent NVDA versions, it is now possible to have these column headers read if you use UIA in Excel (not enabled by default)

**Comment by leonardder** [comment]
> 1. @leonardder the title reads "Excel 2016: column and row headers not read when added using create table feature", but in the description, you only speak of column headers. Can you confirm that row headers are out of the scope of this issue and that there is no way to create native row headers with Excel?

I think this is the case, yet, so I will update the title accordingly.

> 2. In 2023.1 and some other recent NVDA versions, it is now possible to have these column headers read if you use UIA in Excel (not enabled by default)

Yes, I can confirm this and I consider that an acceptable workaround. Lets close this.

**Comment by CyrilleB79** [comment]
@leonardder, it's worth noting that UIA for Excel is not enabled by default in NVDA. I would keep this issue open at least until UIA is enabled by default in NVDA.

**Comment by Adriani90** [comment]
Actually I also think UIA has still alot of bugs in Excel and in MS Office in general, Excel is not really usable with UIA and NVDA yet, the downsides are still to high. Actually we should even reconsider using UIA in Microsoft word by default, using this in a professional environment is still not stable enough. In my view it will still take many years until UIA is as stable as the usual APIs in order to use it alone.

I am reopening this issue.

**Comment by alexandraterec** [comment]
> `Can you confirm that row headers are out of the scope of this issue and that there is no way to create native row headers with Excel?`

@CyrilleB79 I just wanted to confirm that row headers can be added for an Excel table, but this must be manually done after creating a table. Here are the steps to assign row headers, continuing from step 5 of the original message:
6. Check that focus is on any cell within the table
7. Press Alt for the Ribbon and then J, T, C to open the "Table Design" tab and check the "First Column" checkbox in the "Table Style Options" group

Note that when we format data as a table, we automatically get column headers, and only column headers:
* if I leave "My table has headers" checked I only get column headers and only the "Header Row" checkbox is checked.
* if I uncheck "My table has headers"  I still get column headers and under the "Table Design" tab, the "Header Row" checkbox is checked.

Although we've added native column and row headers, NVDA does not read them when moving through the table cells. JAWS 2023 announces both correctly.  cc: @jscholes

**Comment by CyrilleB79** [comment]
> > Can you confirm that row headers are out of the scope of this issue and that there is no way to create native row headers with Excel?
> 
> @CyrilleB79 I just wanted to confirm that row headers can be added for an Excel table, but this must be manually done after creating a table. Here are the steps to assign row headers, continuing from step 5 of the original message: 6. Check that focus is on any cell within the table 7. Press Alt for the Ribbon and then J, T, C to open the "Table Design" tab and check the "First Column" checkbox in the "Table Style Options" group

Yes, confirmed on my side and both headers are reported with UIA. Thanks for the explanation.

To summarize, native definition of headers (columns and rows) can be reported if "Use UI automation to access Microsoft Excel spreadsheet controls when available" is checked in advanced settings. When using legacy mode they are not reported.
